browser/brave_rewards/add_funds_popup.cc

```diff
@@ -0,0 +1,474 @@
+#include "components/content_settings/core/browser/host_content_settings_map.h"
+#include "components/guest_view/browser/guest_view_base.h"
+#include "content/public/browser/page_navigator.h"
+#include "content/public/browser/web_contents.h"
+#include "content/public/browser/web_contents_delegate.h"
+#include "content/public/common/referrer.h"
+#include "net/base/escape.h"
+#include "third_party/blink/public/platform/web_referrer_policy.h"
+#include "ui/base/page_transition_types.h"
+#include "ui/display/display.h"
+#include "ui/display/screen.h"
+#include "ui/views/widget/widget.h"
+
+using content::WebContents;
+
+namespace {
+
+  constexpr int kPopupPreferredHeight = 800;
+  constexpr int kPopupPreferredWidth = 1100;
+
+  // Referrer.
+  const char kRewardsHost[] = "rewards";
+
+  // URL to open in the popup.
+  const char kAddFundsUrl[] = "https://uphold-widget.herokuapp.com/index.php";
+
+  // Content permission URLs and patterns.
+  const char kBraveHost[] = "https://uphold-widget.herokuapp.com/";
+  const char kFirstParty[] = "https://firstParty";
+
+  const char kUpholdPattern[] = "https://[*.]uphold.com/*";
+  const char kNetverifyPattern[] = "https://[*.]netverify.com/*";
+  const char kTypekitPattern[] = "https://use.typekit.net/*";
+
+  const std::map<std::string, std::string> kCurrencyToNetworkMap{
+      {"BTC", "bitcoin"},
+      {"BAT", "ethereum"},
+      {"ETH", "ethereum"},
+      {"LTC", "litecoin"} };
+
+} // namepsace
+
+namespace brave_rewards {
+
+AddFundsPopup::AddFundsPopup()
+    : add_funds_popup_(nullptr),
+      profile_(nullptr),
+      rewards_service_(nullptr),
+      allowed_scripts(false) {}
+
+AddFundsPopup::~AddFundsPopup() {
+  ClosePopup();
+}
+
+// Show existing or open a new popup.
+void AddFundsPopup::ShowPopup(content::WebContents* initiator,
+  RewardsService* rewards_service) {
+  if (add_funds_popup_) {
+    Focus();
+  } else
+    OpenPopup(initiator, rewards_service);
+}
+
+void AddFundsPopup::OpenPopup(content::WebContents* initiator,
+  RewardsService* rewards_service) {
+  DCHECK(!add_funds_popup_);
+  DCHECK(initiator);
+  DCHECK(rewards_service);
+  if (!initiator || !rewards_service)
+    return;
+
+  const std::map<std::string, std::string> addresses = 
+    rewards_service->GetAddresses();
+  if (addresses.empty())
+    return; 
+
+  content::WebContentsDelegate* wc_delegate = initiator->GetDelegate();
+  if (!wc_delegate)
+    return;
+
+  const GURL gurl(kAddFundsUrl);
+  const GURL referrer_gurl(std::string(brave_toolbar::kInternalUIScheme) +
+    kRewardsHost);
+  const content::Referrer referrer(referrer_gurl,
+    blink::kWebReferrerPolicyAlways);
+  content::OpenURLParams params(gurl, referrer,
+    WindowOpenDisposition::NEW_POPUP,
+    ui::PAGE_TRANSITION_LINK, true);
+
+  // Supply addresses via post data. The data is currently in the query string
+  // format (application/x-www-form-urlencoded):
+  // addresses=UrlEscapedBase64EncodedStringifiedJSON.
+  params.uses_post = true;
```

Comment:
per comment in https://github.com/brave/uphold-widget/pull/1#discussion_r233663379, i think this would be better as a GET

Reply:
Can do. I was initially testing this with a GET and query string, so it will definitely work.